netket/optimizer/qgt/qgt_jacobian_pytree.py

```diff
@@ -66,13 +69,22 @@ def QGTJacobianPyTree(
             **kwargs,
         )
 
+    from netket.vqs.exact import ExactState
+
+    if isinstance(vstate, ExactState):
+        samples = split_array_mpi(vstate._all_states)
+        pdf = split_array_mpi(vstate.probability_distribution())
+    else:
+        samples = vstate.samples
+        pdf = None
```

Comment:
This check seems misplaced here. If the QGT (or other objects unrelated to the VQS submodule) needs to distinguish `ExactState` from others, then the variational state interface does not serve its purpose of abstracting away such differences.

Ideally, there should be a method in `VariationalState` that returns `(samples, probabilities)` (or `(samples, None)` for MC, where the samples themselves follow the desired distribution without any weights). Maybe this should even be what is returned by `.samples`. Since this is not backwards-compatible, we unfortunately would need another method. Maybe `samples_with_distribution` or so?

Reply:
This is yet another case of dispatch problems. I knew they'd arise one day....
We need to have some custom code to construct QGTJacobianPyTree for ExactState.

In theory solved by `@dispatch` the constructors but I wanted to avoid it because then *ahem* people *ahem* complain with me.
The method you propose seems like a nice addition, but is it worth adding it just for this use case?

Let's agree on a design on slack..

Reply:
Alright, @PhilipVinc and I discussed this and agreed that, while a bit of a workaround, this solution is fine for the time being.

One more thing I'd like to see is a check in `ExactState.quantum_geometric_tensor` that complains if something other than `QGTJacobianPyTree` is passed (because a QGT using just `samples` would silently give incorrect results, if I am not mistaken).